### ISSUE TITLE
More flexible InteractionTrack.

### ISSUE THREAD

**Issue by MalteThodberg** [ComputationalRegulatoryGenomicsICL/GenomicInteractions-old#35]
As far as I can tell, InteractionTrack is the only way of plotting arches with Gviz. 

I like the plots InteractionTrack produces a lot, but the interface is not very flexible. A few simple changes would make it much more user-friendly:

- Update the documentation of InteractionTrack-class to describe all displayPars.
- Make InteractionTrack work on GInteractions-objects as well as GenomicInteractions-object
- Allow the InteractionTrack constructor function to pass arguments directly, rather than having to specify them later using displayPars (using something like passing ...).
- It should be possible to map height to other columns than "counts, "p.value" and "fdr" (right now I'm overwriting my own variables to be called "counts", which seems somewhat hacky.)

**Comment by mgperry** [comment]
Hi Malte, those are definitely good suggestions, my recollection is that we looked into this at the time and it would have been a lot of effort because of the way GViz worked and it was enough of a hack as it was (I didn't write that bit of the code, so I'm not sure). I don't think there's much chance of us making these changes ourselves since all of us who wrote the package have left the group, although @liz-is still works on Hi-C data.

If you're interested in implementing these changes yourself, I'm happy to help you with any questions you have, and to make sure these changes are pushed back into BioC with the proper acknowledgements etc.

**Comment by MalteThodberg** [comment]
Thanks for the reply!

I had a look at the code, and you are right that it seems to be quite a bit of work to change it (and I'm not very good at coding plots outside of ggplot2).

I'm seriously considering using InteractionTrack in the next update for the CAGEfightR package to plot enhancer-promoter correlations. It doesn't always make sense have arches scale by p-values, and it feels a bit hacky to save something else into the p.value slot.

Come to think of it, maybe such a generic plotting function should really be living in either InteractionSet or Gviz. Then other packages could simply perform slight modifications.

**Comment by mgperry** [comment]
Well, the idea with InteractionSet was that it would be low-level stuff (unlike this package), but Gviz I reckon is the proper home for this, since it's not really possible to extend Gviz without changing the source. It may even have this kind of functionality by now, or it's possible they might be persuaded to add it.

**Comment by mgperry** [comment]
Hey, just checked out your profile and I was in Copenhagen a few weeks ago chatting to Robin Anderssen, quite a coincidence haha.

**Comment by liz-is** [comment]
Hi, 
Thanks for your interest in GenomicInteractions and apologies for not getting to this sooner.

These are mostly things we have considered in the past  and/or would also like to do! Let me go through them:

> Update the documentation of InteractionTrack-class to describe all displayPars.

I agree this would be nice to have, but it would be hard to maintain since the other displayPars are defined in Gviz. We chose to only describe the displayPars we define, to avoid getting out of sync with Gviz and having out of date documentation.

> Make InteractionTrack work on GInteractions-objects as well as GenomicInteractions-object

Right now, if I remember correctly, InteractionTrack depends on having at least one metadata column, while GInteractions object don't have mcols. Perhaps one option would be to create a temporary internal GenomicInteractions object from an input GInteractions object, at least to save the user having to do it. 

> Allow the InteractionTrack constructor function to pass arguments directly, rather than having to specify them later using displayPars (using something like passing ...).

This is how Gviz is implemented - we are following their conventions. It would be confusing otherwise IMO.

> It should be possible to map height to other columns than "counts, "p.value" and "fdr" (right now I'm overwriting my own variables to be called "counts", which seems somewhat hacky.)

Yeah, I have to do this too and it is a bit hacky and suboptimal! This would however require some rewriting of the internal code of InteractionTrack, which is, as you've seen, fairly complex. I'd like to do this at some point but I don't have a lot of time available to work on the package at the moment.

On a related note, I already have some code to make mapping interaction colours to metadata columns much more flexible. This is a good reminder to actually contribute that back to the package (when I get a chance to work on it...). 

I agree with @mgperry that ideally Gviz would be the right home for a general interaction track plotting function. As far as I know this is not yet present in Gviz - don't know if the developers could be persuaded to add it, but I'd probably want to tidy up the code here a bit before suggesting it to them anyway.

**Comment by MalteThodberg** [comment]
@mgperry Yes, I think we only met briefly at the coffee machine, I unfortunately missed you talk.

@liz-is: Thank you for you detailed response, some additional comments:

With regards to implementation of and documentation on InteractionTrack: Most other Gviz functions allow for passing arguments through ... (e.g. DataTrack(data, ...)). For InteractionTrack it only takes "name" in addition to the core data, and then additional settings have to be set afterwards with displayPars() <-. It would be much easier (and more in line with other Gviz functions) if parameters could be passed directly to interaction track like InteractionTrack(gi, name="x", col.interactions="red", col.anchors.fill ="blue", col.anchors.line = "black", etc). Then all the InteractionTrack specific parameter should be documented in InteractionTrack and InteractionTrack-class.

I think there could be a demand for a general arch-plotting functionality in Bioconductor. The only two other packages are ggbio which seem to be unstable, and sushi which is not very extendable.
